BeVolt/BSP/Simulator/Src/BSP_CAN.c

```diff
@@ -1,41 +1,78 @@
-#include "BSP_CAN.h"
-
-/**
- * @brief   Initializes the CAN module that communicates with the rest of the electrical system.
- * @param   None
- * @return  None
- */
-void BSP_CAN_Init(void) {
-
-    // TODO: Initialize the CAN port that connects to the rest of the car.
-}
-
-/**
- * @brief   Transmits the data onto the CAN bus with the specified id
- * @param   id : Message of ID. Also indicates the priority of message. The lower the value, the higher the priority.
- * @param   data : data to be transmitted. The max is 8 bytes.
- * @param   length : num of bytes of data to be transmitted. This must be <= 8 bytes or else the rest of the message is dropped.
- * @return  0 if module was unable to transmit the data onto the CAN bus. Any other value indicates data was transmitted.
- */
-uint8_t BSP_CAN_Write(uint32_t id, uint8_t data[8], uint8_t length) {
-    
-    // TODO: Transmit the data onto the CAN bus with the specified ID.
-
-    return 0;
-}
-
-/**
- * @brief   Gets the data that was received from the CAN bus.
- * @note    Non-blocking statement
- * @pre     The data parameter must be at least 8 bytes or hardfault may occur.
- * @param   id : pointer to store id of the message that was received.
- * @param   data : pointer to store data that was received. Must be 8bytes or bigger.
- * @return  0 if nothing was received so ignore id and data that was received. Any other value indicates data was received and stored.
- */
-uint8_t BSP_CAN_Read(uint32_t *id, uint8_t *data) {
-
-    // TODO: Check if a message has been received.
-    //      If message was received, then store the id and data of the message into the pointers and return a value other than 0.
-
-    return 0;
-}
+#include "BSP_CAN.h"
+#include <stdint.h>
+#include <stdlib.h>
+
+static const char* file = "BSP/Simulator/DataGeneration/Data/CAN.csv";
+
+
+
+/**
+ * @brief   Initializes the CAN module that communicates with the rest of the electrical system.
+ * @param   None
+ * @return  None
+ */
+
+
+
+void BSP_CAN_Init(void) {
+
+    // TODO: Initialize the CAN port that connects to the rest of the car.
+    FILE* fp = fopen(file, "w"); 
+    fclose(fp);
+}
+
+
+
+/**
+ * @brief   Transmits the data onto the CAN bus with the specified id
+ * @param   id : Message of ID. Also indicates the priority of message. The lower the value, the higher the priority.
+ * @param   data : data to be transmitted. The max is 8 bytes.
+ * @param   length : num of bytes of data to be transmitted. This must be <= 8 bytes or else the rest of the message is dropped.
+ * @return  0 if module was unable to transmit the data onto the CAN bus. Any other value indicates data was transmitted.
+ */
+
+uint8_t BSP_CAN_Write(uint32_t id, uint8_t data[8], uint8_t length) {
+    uint32_t offset;
+    // TODO: Transmit the data onto the CAN bus with the specified ID.
+    
+
+    FILE* fp = fopen(file, "w");
+    fprintf(fp, "0x%.3x,", id);
+    for(int i = 0; i < 8; i++){

```

Comment:
In this for loop, you can check (i < length) instead of (i < 8). It is probably best to check both of them, in case someone passes a length greater than 8, so you do not go out of bounds on the data array

Reply:
It seems that .csv files do not have an EOF, so all 8 locations of the data[] parameter for CAN_Read() get filled even if < 8 bytes are printed in CAN_Write()

Reply:
It is fine if the extra locations of data[] get filled with garbage values (this is probably what happens on the microcontroller) since we can tell which locations are valid based on the CAN id. If trying to change this causes a seg fault when CAN_Read() is called (trying to read past end of file), I would just leave it like this.